onnxscript/rewriter/generic_pattern.py

```diff
@@ -1062,46 +1088,68 @@ class OnnxGenericPattern(GenericPattern):
     :param match_proto: the onnx function defining the matching pattern
     :param apply_proto: the onnx function defining the new pattern
     :param validate_mapping: the function used to validate a pattern
+    :param use_onnxscript: tells if the apply_proto is an onnxscript function or
+        a regular function returning a FunctionProto
+    :param opsets: opset to consider when converting the function into ONNX,
+        if not specified, it is opset 18 for the main opset, and opset 1
+        for domain com.microsoft.
     :param verbose: in [0, 10], increase the verbosity to understand why a pattern
         does not match
     """
 
     def __init__(
         self,
         match_proto: onnx.FunctionProto,
-        apply_proto: onnx.FunctionProto,
+        apply_proto: onnx.FunctionProto | typing.Callable,
         validate_mapping: typing.Callable,
+        use_onnxscript: bool = True,
+        opsets: dict[str, "onnxscript.Opset"] | None = None,  # noqa: F821
         verbose: int = 0,
     ):
         super().__init__(verbose=verbose)
         self.match_proto = match_proto
         self._validate_mapping = validate_mapping
         self.apply_proto = apply_proto
+        self.use_onnxscript = use_onnxscript
+        self.opsets = opsets
         self._cache = {}
 
-    def validate_mapping(
-        self, g: oir.Model, deleted_nodes: list[oir.Node], added_nodes: list[oir.Node]
-    ) -> bool:
+    def validate_mapping(self, g: oir.Model, match_result: PatternMatchResult) -> bool:
         """Evaluates the consistency of the replacements."""
-        return self._validate_mapping(g, deleted_nodes, added_nodes)
+        return self._validate_mapping(g, match_result)
 
     def _build_pattern(
-        self, g: ModelWithGraphStructure, fct: typing.Callable
+        self,
+        g: ModelWithGraphStructure,
+        fct: typing.Callable | None = None,
+        kwargs: dict[str, typing.Any] | None = None,
+        match: bool = True,
     ) -> BuilderWithGraphStructure:
-        if fct == self.match_pattern:
-            key = id(g), "match"
-            if key in self._cache:
-                return self._cache[key]
-            onx = self.match_proto
-        elif fct == self.apply_pattern:
-            key = id(g), "apply"
-            if key in self._cache:
-                return self._cache[key]
-            onx = self.apply_proto
+        del fct
+        if match:
+            key = id(g), match, str(kwargs)
         else:
-            raise AssertionError(
-                f"Function {fct} is not {self.match_pattern} or {self.apply_pattern}."
-            )
+            key = id(g), match, str(kwargs)
+            assert not callable(self.apply_proto) or isinstance(kwargs, dict)
+        if key in self._cache:
+            return self._cache[key]
+
+        if match:
+            onx = self.match_proto
```

Comment:
Avoid abbreviations

Reply:
I'd rather create another PR to do that since this PR was started: https://github.com/microsoft/onnxscript/pull/1369.